docs/design/metadata.md

```diff
@@ -0,0 +1,365 @@
+
+Additionally, for the purpose of indexing, the metadata from multiple
+wheels is collected into a single `variants.json` format file. This file
+can be afterwards fetched by the installers to determine which variants
+are available and how to install the provider plugins required to
+determine which variants are supported. This makes it possible to select
+a specific variant wheel without having to download them all.
+
+
+## The common metadata
+
+### Default priorities
+
+The wheel metadata can specify three lists of default priorities:
+
+1. *Namespace priorities* that define both the list of namespaces
+   that are used by default when installing the package in question,
+   and their initial order. Said order may be altered afterwards by user
+   configuration. This list must be defined and non-empty.
+
+2. *Feature priorities* that override the order of features. The initial
+   order is determined by ordering the namespaces according to their
+   priorities (as indicated in point 1.), then replacing them by all
+   their features as ordered by the provider plugin. Afterwards, all
+   features defined in the feature priority list are moved to
+   the beginning of the resulting list, while the relative order
+   of the remaining items remains the same. Said order may be altered
+   by user configuration. This list is optional.
+
+3. *Property priorities* that override the order of properties.
+   The initial order is determined by ordering the features according
+   to their priorities (as indicated in point 2.), then replacing them
+   by all their properties as ordered by the provider plugin.
+   Afterwards, all properties defined in the feature priority list
+   are moved to the beginning of the resulting list, while the relative
+   order of the remaining items remains the same. Said order may be
+   altered by user configuration. This list is optional.
+
+### Provider information
+
+The wheel metadata also includes the provider metadata dictionary,
+keyed on namespaces. This dictionary is obligatory, and must include
+the metadata for all namespaces that are listed in default namespace
+priorities. The values include the following information:
+
+- The *plugin API* value that specifies how to load plugins. It is
+  obligatory. It follows the same format as object references
+  in the [entry points specification](
+  https://packaging.python.org/en/latest/specifications/entry-points/),
+  that is:
+
+  ```
+  importable.module:ClassName
+  ```
+
+  The plugin will be instantiated by the equivalent of:
+
+  ```python
+  import importable.module
+
+  plugin_instance = importable.module.ClassName()
+  ```
+
+- The list of *plugin requirements* that consists of zero or more
+  [dependency specifiers](
+  https://packaging.python.org/en/latest/specifications/dependency-specifiers/)
+  specifying how the provider should be installed. This is optional, but it is
+  necessary to support installing plugins in an isolated environment.
```

Comment:
Why is it optional? It seems to me like it should be either something in-tree or a dependency specifier - unless we expect there to be some built-in plugin in all installers?

Reply:
That's a very good question. To be honest, I don't think we've really revisited this aspect since we've introduced isolated installs. That said, I have mixed feelings about making it required.

On one hand, plugins obviously aren't going to work in isolated environment unless they're installed. On the other hand, there are valid cases for running plugins in non-isolated environment, e.g. during development.

Here, the "required vs. optional" distinction follows the implementation, i.e. if a "required" field is missing we immediately report an error. I wouldn't want to force someone to put dummy requirements just to stop the tool for failing before `pyproject.toml` is final.

Of course, "built-in plugins" are also an interesting idea, and related to the security issues.

Reply:
> On the other hand, there are valid cases for running plugins in non-isolated environment, e.g. during development.

That is an orthogonal question (and yes, non-isolated envs are definitely important, equally so I'd say). In non-isolated envs, the dependency is stil required. And hence should still be specified, and may be enforced by build frontends. E.g., see how `pypa/build` still does dependency verification for all specified deps when you run it with `python -m build --no-isolation`.

Reply:
> Of course, "built-in plugins" are also an interesting idea, and related to the security issues.

Yes, that is interesting indeed. But that's for the tool that does the building in. E.g., if `uv` decided to provide `x86-64` internally, it must know that it can skip installing the plugin for that when it sees it in the dependency specifier list.